src/libraries/System.Private.CoreLib/src/System/SpanHelpers.T.cs

```diff
@@ -939,14 +985,55 @@ public static void Fill<T>(ref T refData, nuint numElements, T value)
             if (valueLength == 0)
                 return -1;  // A zero-length set of values is always treated as "not found".
 
-            int index = -1;
-            for (int i = 0; i < valueLength; i++)
+            // See comments in IndexOfAny(ref T, int, ref T, int) above regarding algorithmic complexity concerns.
+            // This logic is similar, but it runs backward.
+
+            if (typeof(T).IsValueType)
             {
-                int tempIndex = LastIndexOf(ref searchSpace, Unsafe.Add(ref value, i), searchSpaceLength);
-                if (tempIndex > index)
-                    index = tempIndex;
+                for (int i = searchSpaceLength - 1; i >= 0; i--) // n.b. 'i' is signed so we can use negative bit to determine end of sequence
+                {
+                    T candidate = Unsafe.Add(ref searchSpace, i);
+                    for (int j = 0; j < valueLength; j++)
+                    {
+                        if (Unsafe.Add(ref value, j).Equals(candidate))
+                        {
+                            return i;
+                        }
+                    }
+                }
             }
-            return index;
+            else
+            {
+                // Calling IEquatable<T>.Equals (virtual dispatch). We'll perform the null check
+                // in the outer loop instead of in the inner loop to save some branching.
+
+                for (int i = searchSpaceLength - 1; i >= 0; i--) // n.b. 'i' is signed so we can use negative bit to determine end of sequence
```

Comment:
It's not clear to me what the comment is referring to.  Isn't this a typical "walk backwards" loop?

Reply:
It was more of a "don't be clever and use `nuint` here" warning. But I agree it's inartfully worded, and honestly anybody who tries that will immediately see the failure. So the warning's probably not needed.

Reply:
> It was more of a "don't be clever and use nuint here" warning

Ah :-) Using nuint hadn't even crossed my mind... guess int is just too baked into my conscious :-)